Adding new tasks and moving old ones for 'Femto3D' package

glromane opened the pull request:
--Implementing small changes in "singletrackselector" task and moving them to "Femto3D" folder in PWGCF branch;

--adding a QA task that uses tables from "singletrackselector";

--adding a femto task that does mixing, pair creation for obtaining a femtoscopic correlation function;

victor-gonzalez reviewed (changes requested):
Please, use the same sub-directories structure for `Femto3D` as the other functional areas, as for example `FemtoUniverse`

Have in mind that the mandate is that there should be only one derived data datamodel for `Femto` analysis. There has been extensive work in this area and such a datamodel already exists. We will further iterate on that (see my e-mail)

zchochul commented:
Please try to reduce the amount of commented code. (like here: https://github.com/AliceO2Group/O2Physics/blob/08eb0e5793909714d354b60b31281572316af107/PWGCF/Femto3D/femto3dPairTask.cxx#L45)

victor-gonzalez commented on PWGCF/Femto3D/TableProducer/singleTrackSelector.cxx:
```diff
@@ -0,0 +1,180 @@
+        // if satisfied, want to continue in the upper loop (over tracks) -- skip the current track
+        // cannot use simple 'continue' since it will be applied to the current loop, so have to use 'goto'
+        if (TOFselection(track, std::make_pair(i, rejectWithinNsigmaTOF))) {
+          goto skip;
```
Please, no `goto`
Use meaningful name variables to control the execution flow

glromane replied:
I put a flag instead. No 'goto' anymore.

victor-gonzalez commented on PWGCF/Femto3D/TableProducer/singleTrackSelector.cxx:
```diff
@@ -0,0 +1,180 @@
+          break; // break the loop with particlesToKeep after the 'if' condition is satisfied -- don't want double entries
+        }
+
+    skip:; // here we send with 'goto'
```
Please, no labels
Use meaningful name variables to control the execution flow

glromane replied:
Removed.

victor-gonzalez commented on PWGCF/Femto3D/Tasks/femto3dPairTask.cxx:
```diff
@@ -0,0 +1,305 @@
+      for (auto& track : tracks.sliceBy(perCollId, collision.index())) {
+
+        if (track.sign() == _sign_1 && (track.p() < _PIDtrshld_1 ? TPCselection(track, TPCcuts_1) : TOFselection(track, TOFcuts_1))) { // filling the map: eventID <-> selected particles1
+          Particle = new FemtoParticle(track.energy(particle_mass(_particlePDG_1)), track.px(), track.py(), track.pz(), track.eta(), track.phi());
+          Particle->SetSign(_sign_1);
+          Particle->SetMagField(collision.magField());
+          selectedtracks_1[track.singleCollSelId()].push_back(Particle);
+
+          registry.fill(HIST("p_first"), track.p());
+          if (_particlePDG_1 == 2212) {
+            registry.fill(HIST("nsigmaTOF_first"), track.p(), track.tofNSigmaPr());
+            registry.fill(HIST("nsigmaTPC_first"), track.p(), track.tpcNSigmaPr());
+          }
+          if (_particlePDG_1 == 1000010020) {
+            registry.fill(HIST("nsigmaTOF_first"), track.p(), track.tofNSigmaDe());
+            registry.fill(HIST("nsigmaTPC_first"), track.p(), track.tpcNSigmaDe());
+          }
```
You are replicating in memory everything that you get from the tables with a much more inefficient way of accessing to the stored information. 
This will have a huge penalty in execution time and in memory requirements
This goes against the O2 execution model (SoA vs AoS)
It seems that one of the reasons for storing the information is for extracting the mixed event results
You have to use the mixed events framework that is standard in O2Physics
It works without any kind of data replication

glromane replied:
Maybe it doesn't seem right but I'll try to explain why I think it's more convenient to use it like this. The main reason is the mixing (I put a more detailed comment about the mixing in the main reply). Since we need to store some info per particle for the mixing later I decided to use this "FemtoParticle" object that contains the bare minimum: pT, Eta, Phi, magField and signedPDG (updated in the last commit wrt to the previous one -- now consumes less memory).

Actually I was trying to store iterators/pointers to the tracks but the ones that are created after slicing or filtering are transient (temporary) ones and cannot be used to access an object (for the mixing) after the selection. So, I don't see any other way to do it apart from creating another additional table which would lead to duplication anyway. But using the class "FemtoPatricle" is more convenient because I can put there needed functions, pass the object to the "FemtoPair" etc.

victor-gonzalez commented on PWGCF/Femto3D/Tasks/femto3dPairTask.cxx:
```diff
@@ -0,0 +1,305 @@
+        if (IsIdentical || track.sign() != _sign_2)
+          continue;
+        else if (!TOFselection(track, std::make_pair(_particlePDGtoReject, _rejectWithinNsigmaTOF)) && (track.p() < _PIDtrshld_2 ? TPCselection(track, TPCcuts_2) : TOFselection(track, TOFcuts_2))) { // filling the map: eventID <-> selected particles2 if (see condition above ^)
+          Particle = new FemtoParticle(track.energy(particle_mass(_particlePDG_2)), track.px(), track.py(), track.pz(), track.eta(), track.phi());
+          Particle->SetSign(_sign_2);
+          Particle->SetMagField(collision.magField());
+          selectedtracks_2[track.singleCollSelId()].push_back(Particle);
+
+          registry.fill(HIST("p_second"), track.p());
+          if (_particlePDG_2 == 2212) {
+            registry.fill(HIST("nsigmaTOF_second"), track.p(), track.tofNSigmaPr());
+            registry.fill(HIST("nsigmaTPC_second"), track.p(), track.tpcNSigmaPr());
+          }
+          if (_particlePDG_2 == 1000010020) {
+            registry.fill(HIST("nsigmaTOF_second"), track.p(), track.tofNSigmaDe());
+            registry.fill(HIST("nsigmaTPC_second"), track.p(), track.tpcNSigmaDe());
+          }
```
Same comment as above

victor-gonzalez commented on PWGCF/Femto3D/Tasks/femto3dPairTask.cxx:
```diff
@@ -0,0 +1,305 @@
+
+    else { //====================================== mixing non-identical ======================================
+
+      for (auto i = selectedtracks_1.begin(); i != selectedtracks_1.end(); i++) { // start of same event non-identical
+
+        auto j = selectedtracks_2.find(i->first);
+        if (j == selectedtracks_2.end())
+          continue;
+
+        for (auto ii : i->second) {
+          for (auto iii : j->second) {
+            FemtoPair* Pair = new FemtoPair(ii, iii, IsIdentical);
+            if (!Pair->IsClosePair(_deta, _dphi, _radiusTPC))
+              registry.fill(HIST("SE"), Pair->GetKstar()); // close pair separation doesn't work properly, need eta and phi at some position in the barrel
+          }
+        }
+      } // end of same event non-identical
+
+      for (auto i = mixbins.begin(); i != mixbins.end(); i++) { // start of mixed event non-identical
+        if ((i->second).size() < 2)
+          continue;
+
+        std::vector<std::vector<int>> comb_idx;
+        comb(comb_idx, (i->second).size(), 2);
+
+        for (auto indx : comb_idx) {
+          int colId_1 = (i->second)[indx[0]];
+          int colId_2 = (i->second)[indx[1]];
+
+          for (auto ii : selectedtracks_1[colId_1]) {
+            for (auto iii : selectedtracks_2[colId_2]) {
+              FemtoPair* Pair = new FemtoPair(ii, iii, IsIdentical);
+              if (!Pair->IsClosePair(_deta, _dphi, _radiusTPC))
+                registry.fill(HIST("ME"), Pair->GetKstar());
+            }
+          }
+        }
+      } // end of mixed event non-idetical
+
+    } //====================================== end of mixing non-identical ======================================
```
Same comment as above

glromane replied:
Here you're right. I fixed this and now I use only one "FemtoPair" object and just change the stored objects inside it. Thanks for pointing it out to me.

victor-gonzalez reviewed (changes requested):
Thanks for reallocating the source files in the suggested structure

For the actual code, please, consider the suggested comments, specially the one encouraging the use of the standard O2Physics mixed events framework and not to replicate in memory what is received from tables

glromane commented:
> Thanks for reallocating the source files in the suggested structure
> 
> For the actual code, please, consider the suggested comments, specially the one encouraging the use of the standard O2Physics mixed events framework and not to replicate in memory what is received from tables

Dear Victor, thanks for your comments, here is my overall reply (I also commented on your separate comments).

I took into account your suggestions and applied them in the new commit.
Small things as namespace, goto etc. I fixed. For the "FemtoParticle", "FemtoPair" objects and mixing procedure I'm putting a more detailed comment:

First, for the tables I cannot use filtering for the PID selection which means that even in the filtered with the track cuts table I will have lots of unwanted entries. And if I pass the table with the "fake" tracks to the mixer (as in the FemtoDream) I'll have lots of "fake" pairs to process the PID selection for. The number of pairs scales as n*(n-1) (x2 because it's a pair) and I think it's very inefficient and instead I do the PID before the mixing and pairs I build doesn't need to be checked anymore. So, instead of 2*n*(n-1) PID checks, I do it only n-times. Yes, for that I need to store the info and allocate memory but I gain a lot in CPU efficiency. And for the use of O2 mixing algorithm after the PID I'd need to make another table and duplicate the objects anyway, so it's the same only use of the "FemtoParticle" class is much more convenient for me.

There are also some other for wrt to the previous commit made for the sake of efficiency (for example I got rid of the function that does the permutations and replaced it with just nested loops).

We will continue working on improvements and of course, if we don't meet the hyper loop requirements we will deal with it, but before we need to have our tasks there to test and see how to proceed.

If there is something unclear, please, tell me, I'll give you more details. Thank you.

Sincerely yours, Gleb.

glromane commented on PWGCF/Femto3D/Core/femto3dPairTask.h:
```diff
@@ -0,0 +1,224 @@
+    TVector3 qinv = first4momentum->Vect() - second4momentum->Vect();
+    return 0.5 * abs(qinv.Mag());
+  }
+}
```
Fixed, put them in the "o2::aod::singletrackselector::" namespace.

glromane commented on PWGCF/Femto3D/DataModel/singletrackselector.h:
```diff
@@ -0,0 +1,201 @@
+} // namespace singletrackselector
+
+DECLARE_SOA_TABLE_FULL(SingleTrackSel, "SelTracks", "AOD", "STSEL", // Table of the variables for single track selection.
+                       o2::soa::Index<>,
```
Corrected.

glromane commented on PWGCF/Femto3D/DataModel/singletrackselector.h:
```diff
@@ -0,0 +1,201 @@
+
+template <typename TrackType>
+inline bool TPCselection(TrackType const& track, std::pair<int, std::vector<float>> const& PIDcuts)
+{
```
This structure contains a PDG/PIDcuts pair and if I'm not mistaken the pair is created in only ones at the beginning of a task (QA of PairTask) and this function only reads it since it's by a "const" reference.

victor-gonzalez commented:
We cannot approve the code as it is
One of the rules we have is to approve code that could be use as reference for other people 
Clearly this code is against the O2 rules

Consider this procedure. Once you have your filtered collisions/tracks, your derived data, consider using an additional column you can join to your filtered tracks table which will tag your filtered tracks according to your selection criteria; then use the standard O2 mixed events after filtering the filtered tracks over the additional column

If you plan to elaborate more on your code exclude everything which is not the filter
I will approve the filter.  Then, running the filter on hyperloop will allow you to get enough derived data for your local tests in case you need them

jgrosseo reviewed (changes requested):
See: https://github.com/AliceO2Group/O2Physics/pull/3675#issuecomment-1815138879

jgrosseo commented:
Just to add: if you think something is missing in the O2 event mixing framework, please let us know. We can organize a short slot in the WP4/14 meeting and it can be discussed if it cannot be resolved by simply replying here...
